### PR TITLE
Strict input validation

### PR DESCRIPTION
**Pull request by AnandAili**
Added playground status and manager role related validations

### CODE REVIEW

File: technical/validations.md [hunk 1]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
 ### Playground
-- The name of the playground is between 3 en 40 characters
-- Two playgrounds cannot be located within .. metres of each other
-- The maximum number of playgrounds that can exist within the system is ..
+- the name of the playground is between 3 en 40 characters
+- two playgrounds can not have the same name
+- no whitespaces in the beginning and trailing of the playground name
+- two playgrounds cannot be located within 100 metres of each other
+- the maximum number of playgrounds that can exist within the system is 1000
+- at least 4 decimal precisions required for longitude and latitude of the playground location, in order to avoid duplicate entries
```
Comment (dimitriirtimid) [inline technical/validations.md]: this one can go

File: technical/validations.md [hunk 2]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- a user can join the playground where he is already a volunteer in, then system just returns success and ignore the request
```
Comment (dimitriirtimid) [inline technical/validations.md]: ok, but should it a general design for aggrates/command handler to ignore commands that will not change the state of the system

File: technical/validations.md [hunk 3]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- the id, name, lat, lng and status are required properties during creation of a new playground
+- for onboarding smokefree playground, the smokeFreeDate must be past
```
Comment (dimitriirtimid) [inline technical/validations.md]: onboarding smokefree playground -> onboarding playground with status 'finished' (that claims to be smokefree)

File: technical/validations.md [hunk 4]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- smokefreeDate can only be reset by managers, if the smokefreeDate is passed then it can not be reset
```
Comment (dimitriirtimid) [inline technical/validations.md]: remove: if the smokefreeDate is passed then it can not be reset

File: technical/validations.md [hunk 5]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- the status can not be stopped unless the playground managers choose that the playground can not be smokefree
+- the type can not be smokefree unless it get consecutive 500 smokefree playground observations and status must be finished
```
Comment (dimitriirtimid) [inline technical/validations.md]: nice idea, leave it in as TBD. We need to think when to assign which status

File: technical/validations.md [hunk 6]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- the status can not be finished unless all the phases and actions are completed
```
Comment (dimitriirtimid) [inline technical/validations.md]: it actually, the actions are not obligatory. We may a frontend validation on this. Also note that some smokefree playground will be put in the system in hindsight.

File: technical/validations.md [hunk 7]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
 
 ### Playground workspace
-- The maximum number of volunteers that can enter a playground workspace is 200
-- The maximum number of managers in a workspace is ..
+- the maximum number of volunteers that can enter a playground workspace is 200
+- the maximum number of managers in a workspace is volunteers/5
```
Comment (dimitriirtimid) [inline technical/validations.md]: max 3

File: technical/validations.md [hunk 8]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- only playground managers can decide whether playground can be smokefree or not
+- the playground can not be decided if status is not_started or stopped
```
Comment (dimitriirtimid) [inline technical/validations.md]: ... decided to be smokefree

File: technical/validations.md [hunk 9]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- only playground managers can set smokefreeDate and they cannot set smokefreeDate unless they decide playground to be smokefree
```
Comment (dimitriirtimid) [inline technical/validations.md]: unless they decide playground to be smokefree -> when the smokefreedecision has not been made yet

File: technical/validations.md [hunk 10]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- managers will not be able to spread/share the date unless smokefreeDate is chosen
```
Comment (dimitriirtimid) [inline technical/validations.md]: cant validate this on the server -> drop

File: technical/validations.md [hunk 11]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- the last playground manager can not unsubscribe without choosing one from the recommended list of volunteers
```
Comment (dimitriirtimid) [inline technical/validations.md]: drop this one

File: technical/validations.md [hunk 12]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- if a volunteer  claims the manager role where he is already a manager , then system return success by ignoring the request
```
Comment (dimitriirtimid) [inline technical/validations.md]: ok, but dont spec as separate validation

File: technical/validations.md [hunk 13]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+
 
 
 ### Chat message
 - a chat message cannot consist of whitespace only
 - the minimum length of a chat message is 1 characters
 - the maximum length of a chat message is 300 characters
```
Comment (dimitriirtimid) [inline technical/validations.md]: 3000

File: technical/validations.md [hunk 14]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
-- this is the set of valid characters for a chat message: ...  (or provide a regexp)
+- this is the set of valid characters for a chat message: [!@&(),.?":]
+
+### Playground Observation
+- a comment message can't consist of whitespaces only
+- no whitespaces in the beginning or/and trailing of a comment
+- the minimum length of a comment message is 100 characters
```
Comment (dimitriirtimid) [inline technical/validations.md]: 1

File: technical/validations.md [hunk 15]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- the maximum length of a comment message is 2000 characters
+- how frequently a volunteer can indicate his observations about the playground
```
Comment (dimitriirtimid) [inline technical/validations.md]: once a day (per volunteer)

File: technical/validations.md [hunk 16]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- the indicated playground observation will be shown after verification of the comment by anyone of the managers
```
Comment (dimitriirtimid) [inline technical/validations.md]: not needed, but in future we need to a ' report abuse' button that anyone can press

File: technical/validations.md [hunk 17]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- consecutive 500 smokefree playground observations considered that The playground is now smokefree
```
Comment (dimitriirtimid) [inline technical/validations.md]: 50

File: technical/validations.md [hunk 18]
```diff
@@ -7,20 +7,49 @@ these validations will be duplicated in the front-end code for an optimal user e
...
+- the maximum number of playground observations for the playground can be stored are 1000
+- the smokefree indication, observer, initiativeId are required and no default value is set
+- multiple volunteers can not indicate observation for a particular playground at the same time
```
Comment (dimitriirtimid) [inline technical/validations.md]: I understand, but just only do the validation of one observation per volunteer per playground per day

**Comment by AnandAili** [comment]
Could you please review it again and merge it to base branch

**Pull request closed without merging**